lib/ImportAssetMetadata.js

```diff
@@ -0,0 +1,157 @@
+const PimCategory = require('../service/PimCategory');
+const PimAttributeValue = require('../service/PimAttributeValue');
+const PimAttributeLabel = require('../service/PimAttributeLabel');
+const { prepareIdsForSOQL } = require('./utility');
+
+class ImportAssetMetadata extends ImportClass {
+  /**
+   * @param {HttpRequest} req
+   * @param {HttpResponse} res
+   */
+  constructor(req, res) {
+    super(req, res);
+
+    this.assetNames = [];
+    this.parentNames = [];
+    this.labelNames = [];
+    this.categoryMap = new Map();
+    this.existingAssetMap = new Map();
+    this.newAssetList = [];
+    this.labelNameIdMap = new Map();
+    this.assetLabelValueMap;
+    this.tempAttrValuesToUpdate = [];
+    this.tempAttrValuesToInsert =[];
+
+    this.start();
+  }
+
+  async start() {
+    await this.connect();
+
+    if (this.propelParser.nodes && this.propelParser.nodes.length > 0) {
+      this.populateAssetNames();
+      this.populateLabelNames();
+      await this.populateLabelNameIdMap();
+      await this.populateCategoryMap();
+      await this.populateExistingAssetMap();
+      await this.populateDigitalAssetAttributeValues();
+      await this.updateExistingAssets();
+      this.populateNewAssetList();
+    }
+
+    // finish up
+    await this.log.sendReport();
+  }
+
+  populateAssetNames() {
+    this.propelParser.nodes.forEach(node => {
+      this.assetNames.push(`'${node.digital_asset_id}'`);
+    });
+  }
+
+  populateLabelNames() {
+    Object.keys(this.propelParser.nodes[0]).forEach((header) => {
+      if (header !== 'digital_asset_id' && header !== 'category_id' && header !== 'cdn_url') {
+        this.labelNames.push(`'${header}'`);
+      }
+    });
+  }
+
+  async populateLabelNameIdMap() {
+    const pimAttributeLabel = new PimAttributeLabel(this.helper, this.log, this.labelNames);
+    await pimAttributeLabel.populate('Label__c');
+    this.labelNameIdMap = pimAttributeLabel.getNameMap();
+  }
+
+  async populateCategoryMap() {
+    const pimCategory = new PimCategory(this.helper, this.log);
+    await pimCategory.populate();
+    this.categoryMap = pimCategory.getIdMap();
+  }
+
+  // get a Map of <Digital Asset Name => Digital_Asset__c Id> for assets that already exist in PIM
+  async populateExistingAssetMap() {
+    const pimAssets = new PimAsset(this.helper, this.log, [
+      ...this.assetNames,
+      ...this.parentNames
+    ]);
+    await pimAssets.populate();
+    this.existingAssetMap = pimAssets.getNameMap();
+  }
+
+  async populateDigitalAssetAttributeValues() {
+    const pimAttributeValue = new PimAttributeValue(this.helper, this.log, []);
+    await pimAttributeValue.populateWithDigitalAssetValues(
+      prepareIdsForSOQL(Array.from(this.existingAssetMap.values())), 
+      this.labelNames
+    );
+    this.assetLabelValueMap = pimAttributeValue.sortAccordingToDigitalAssetAndLabel();
+  }
+
+  async updateExistingAssets() {
+    for (let x = 0; x < this.propelParser.nodes.length; x++) {
```

Comment:
I wrote the below comment first but in general this for loop might be doing to much. You are going through each node and updating or inserting for each node. Instead you want to keep building your collections and submit one update and insert at the end. In your case having two DML operations instead of 2n, n being the number of nodes.